File: python/flux/triton/gemm_rs.py
################################################################################
#
# Copyright 2025 ByteDance Ltd. and/or its affiliates. All rights reserved.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
################################################################################

from typing import List, Optional

import torch
import torch.distributed
import triton
from flux_triton.kernels.gemm_rs import (
    add_continous_kernel,
    gemm_rs_kernel,
    gemm_rs_nvlink_kernel,
    get_tune_config,
    reduce_nvlink_kernel,
)

import flux
from cuda import cuda, cudart


def CUDA_CHECK(err):
    if isinstance(err, cuda.CUresult):
        if err != cuda.CUresult.CUDA_SUCCESS:
            raise RuntimeError(f"Cuda Error: {err}: {cuda.cuGetErrorName(err)}")
    elif isinstance(err, cudart.cudaError_t):
        if err != cudart.cudaError_t.cudaSuccess:
            raise RuntimeError(f"Cuda Error: {err}: {cudart.cudaGetErrorString(err)}")
    else:
        raise RuntimeError(f"Unknown error type: {err}")


class GemmRSTritonPCIe(torch.nn.Module):
    def __init__(
        self,
        pg: torch.distributed.ProcessGroup,
        input_dtype: torch.dtype,
        output_dtype: torch.dtype,
        max_m: int = 8192,
        n: int = 8192,
    ):
        self.pg = pg
        self.rank: int = pg.rank()
        self.world_size: int = pg.size()
        self.max_m: int = max_m
        assert max_m % self.world_size == 0, f"max_m({max_m}) % world_size({self.world_size}) != 0"
        self.n: int = n
        self.input_dtype: torch.dtype = input_dtype
        self.output_dtype: torch.dtype = output_dtype
        self.barriers: List[torch.Tensor] = flux.create_tensor_list(
            (
                3,
                self.world_size,
            ),  # 1st for ready flag, 2nd for ready tile counter, 3rd for reduce flag
            torch.int32,
            self.pg,
        )
        self.barrier = self.barriers[self.rank]
        self.gemm_ready_flags = [barrier[0, :] for barrier in self.barriers]
        self.gemm_tile_counters = [barrier[1, :] for barrier in self.barriers]
        self.reduce_ready_flags = [barrier[2, :] for barrier in self.barriers]
        self.gemm_ready_flag = self.gemm_ready_flags[self.rank]
        self.gemm_tile_counter = self.gemm_tile_counters[self.rank]
        self.reduce_ready_flag = self.reduce_ready_flags[self.rank]

        self.cp_stream: torch.cuda.Stream = torch.cuda.Stream(
            priority=-1
        )  # high priority stream is needed
        self.output_tensors: List[torch.Tensor] = flux.create_tensor_list(
            (max_m, n), self.output_dtype, self.pg
        )
        assert len(self.output_tensors) == self.world_size
        self.output_tensor: torch.Tensor = self.output_tensors[self.rank]
        self.reduce_tensors: list[torch.Tensor] = flux.create_tensor_list(
            (max_m, n), self.output_dtype, self.pg
        )
        assert len(self.reduce_tensors) == self.world_size
        self.reduce_tensor: torch.Tensor = self.reduce_tensors[self.rank]
        self.group_barrier = flux.GroupBarrier(self.pg, False)

    def set_copy_done(self, rank: int, segment: int, stream: torch.cuda.Stream):
        (err,) = cuda.cuStreamWriteValue32(
            stream.cuda_stream,
            self.reduce_ready_flags[rank][segment].data_ptr(),
            1,
            cuda.CUstreamWriteValue_flags.CU_STREAM_WRITE_VALUE_DEFAULT,
        )
        CUDA_CHECK(err)

    def wait_gemm_ready(self, rank: int, segment: int, stream: torch.cuda.Stream):
        (err,) = cuda.cuStreamWaitValue32(
            stream.cuda_stream,
            self.gemm_ready_flags[rank][segment].data_ptr(),
            1,
            cuda.CUstreamWaitValue_flags.CU_STREAM_WAIT_VALUE_EQ,
        )
        CUDA_CHECK(err)

    def wait_copy_ready(self, rank: int, segment: int, stream: torch.cuda.Stream):
        (err,) = cuda.cuStreamWaitValue32(
            stream.cuda_stream,
            self.reduce_ready_flags[rank][segment].data_ptr(),
            1,
            cuda.CUstreamWaitValue_flags.CU_STREAM_WAIT_VALUE_EQ,
        )
        CUDA_CHECK(err)

    def run_reduce_scatter_ring_push_1d(
        self,
        M: int,
        output: Optional[torch.Tensor],
        stream: torch.cuda.Stream,
    ):
        M_per_rank = M // self.world_size
        if output:
            assert output.dtype == self.output_dtype
            M_, N_ = output.shape
            assert M_ == M_per_rank and N_ == self.n and output.is_cuda

        # M_per_rank = M // self.world_size
        to_rank = (self.rank - 1 + self.world_size) % self.world_size
        with torch.cuda.stream(stream):
            for stage in range(self.world_size):
                send_segment = (self.rank + stage + 1) % self.world_size
                dst = self.reduce_tensors[to_rank][
                    send_segment * M_per_rank : (send_segment + 1) * M_per_rank, :
                ]
                src = (
                    output
                    if stage == self.world_size - 1 and output is not None
                    else self.output_tensor[
                        send_segment * M_per_rank : (send_segment + 1) * M_per_rank, :
                    ]
                )
                self.wait_gemm_ready(self.rank, send_segment, stream)
                if stage != 0:
                    self.wait_copy_ready(self.rank, send_segment, stream)
                    buffer = self.reduce_tensors[self.rank][
                        send_segment * M_per_rank : (send_segment + 1) * M_per_rank, :
                    ]
                    # src.add_(buffer)
                    add_continous_kernel[(16,)](
                        src,
                        buffer,
                        src,
                        dst.numel(),
                        num_warps=32,
                        BLOCK_SIZE=1024 * 8 * 4,
                    )
                if stage == self.world_size - 1:
                    break
                dst.copy_(src)
                self.set_copy_done(to_rank, send_segment, stream)

    def forward(
        self,
        x: torch.Tensor,
        weight: torch.Tensor,
        bias: Optional[torch.Tensor] = None,
        output: Optional[torch.Tensor] = None,
        input_scale: Optional[torch.Tensor] = None,
        weight_scale: Optional[torch.Tensor] = None,
        output_scale: Optional[torch.Tensor] = None,  # TODO(wenlei.bao) how to handle output_scale?
        transpose_weight: bool = False,
        fast_accum: bool = False,  # TODO(houqi.1993) how to do triton fast_acc?
    ):
        if not transpose_weight:
            weight = weight.t()

        stream = torch.cuda.current_stream()
        # Check constraints.
        M, K = x.shape
        M_per_rank = M // self.world_size
        assert K == weight.shape[0], f"Incompatible dimensions: {x.shape} vs {weight.shape}"
        assert x.is_contiguous(), "Matrix A must be contiguous"
        assert M % self.world_size == 0, f"{M} % {self.world_size} != 0"
        assert self.max_m >= M, f"{M} > {self.max_m}"
        K, N = weight.shape
        assert self.n == N, f"{self.n} != {N}"
        # 1D launch kernel where each block gets its own program.
        grid = lambda META: (
            triton.cdiv(M, META["BLOCK_SIZE_M"]) * triton.cdiv(N, META["BLOCK_SIZE_N"]),
        )
        self.barrier.zero_()
        self.group_barrier.barrier_all(stream.cuda_stream)

        self.cp_stream.wait_stream(stream)
        segments_info_cpu = torch.zeros(
            (self.world_size * 4,), dtype=torch.int16, device="cpu", pin_memory=True
        )

        config = get_tune_config(
            M, N, K, self.world_size, False, not transpose_weight, self.input_dtype
        )
        TILE_SIZE_M = config["BLOCK_SIZE_M"]
        TILE_SIZE_N = config["BLOCK_SIZE_N"]

        flux.calc_gemm_rs_threadblock_segments_info(
            segments_info_cpu,
            (M, N),
            (TILE_SIZE_M, TILE_SIZE_N),
            self.rank,
            self.world_size,
            self.world_size,
            1,  # nnodes. only nodes==1 supported
            False,  # use 2d ring. not implemented here
            False,  # per-tile flag. not implemented here
        )
        segments_info = torch.zeros_like(segments_info_cpu, device="cuda")
        segments_info.copy_(segments_info_cpu, non_blocking=True)
        is_s8 = self.input_dtype == torch.int8
        is_fp8 = flux.util.is_fp8_dtype(x.dtype)
        # NOTE: strange but necessary. to align with flux. better put it outside
        bias = None if is_s8 and self.rank != 0 else bias
        # run gemm
        gemm_rs_kernel[grid](
            x,
            weight,
            bias,
            self.output_tensor,  #
            input_scale,
            weight_scale,
            segments_info,
            M,
            N,
            K,  #
            x.stride(0),
            x.stride(1),  #
            weight.stride(0),
            weight.stride(1),  #
            self.output_tensor.stride(0),
            self.output_tensor.stride(1),  #
            self.rank,
            self.world_size,
            self.gemm_tile_counter,
            self.gemm_ready_flag,
            BIAS_DIMS=("none" if bias is None else ("n" if is_s8 or is_fp8 else "mn")),
            **config,
        )
        self.run_reduce_scatter_ring_push_1d(M, output, self.cp_stream)
        stream.wait_stream(self.cp_stream)
        # print(self.barrier)
        self.group_barrier.barrier_all(stream.cuda_stream)
        return self.output_tensor[self.rank * M_per_rank : (self.rank + 1) * M_per_rank, :]


class GemmRSTritonNVLink(torch.nn.Module):
    def __init__(
        self,
        pg: torch.distributed.ProcessGroup,
        weight: torch.Tensor,
        max_m: int,
        n: int,
        k: int,
        fused: bool = False,
        dump_ptx: bool = False,
    ):
        self.pg = pg
        self.rank = pg.rank()
        self.world_size = pg.size()
        self.weight = weight
        self.max_m = max_m
        self.n = n
        assert max_m % self.world_size == 0, f"max_m={max_m} % world_size={self.world_size} != 0"
        assert (
            weight.shape[0] * self.world_size == k
        ), f"K={k} != weight.shape[1]={weight.shape[0]}*world_size={self.world_size}"
        assert weight.shape[1] == n
        self.max_m_per_rank = max_m // self.world_size
        self.dtype = weight.dtype
        self.output_dtype = weight.dtype
        self.barriers: List[torch.Tensor] = flux.create_tensor_list(
            (self.world_size, 1), torch.int32, self.pg
        )
        self.barrier_ptrs = torch.tensor([ptr.data_ptr() for ptr in self.barriers]).cuda()
        self.barrier = self.barriers[self.rank]
        self.rs_stream: torch.cuda.Stream = torch.cuda.Stream()
        self.output_tensors: List[torch.Tensor] = flux.create_tensor_list(
            (self.max_m, n), self.output_dtype, self.pg
        )
        self.output_tensor_ptrs = torch.tensor(
            [buffer.data_ptr() for buffer in self.output_tensors]
        ).cuda()
        assert len(self.output_tensors) == self.world_size
        self.output_tensor: torch.Tensor = self.output_tensors[self.rank]
        self.output = torch.zeros(
            [self.max_m_per_rank, self.n], dtype=self.output_tensor.dtype
        ).cuda()
        # fusion control
        self.fused = fused
        # debug control
        self.dump_ptx = dump_ptx
        self.group_barrier = flux.GroupBarrier(self.pg, False)

    def forward(self, x: torch.Tensor):
        stream = torch.cuda.current_stream()
        assert x.shape[1] == self.weight.shape[0]
        assert x.is_contiguous(), "Matrix A must be contiguous"
        M, K_per_rank = x.shape
        K = K_per_rank * self.world_size
        _, N = self.weight.shape
        assert self.max_m >= M, "max_m < M"

        self.barrier.zero_()
        self.group_barrier.barrier_all(stream.cuda_stream)

        def ceil(a, b):
            return (a + b - 1) // b

        m_per_rank = ceil(M, self.world_size)
        blocks_m_per_rank = ceil(m_per_rank, 128)
        gemm_blocks = ceil(M, 128) * ceil(N, 128)
        reduce_blocks = ceil(m_per_rank, 128) * ceil(N, 128)

        if self.fused:
            raise NotImplementedError()

        else:
            # GEMM RS hyper parameters
            GEMM_BLOCK_M = 128
            GEMM_BLOCK_N = 128
            GEMM_BLOCK_K = 32
            GEMM_GROUP_M_SIZE = 8
            REDUCE_BLOCK_M = 128
            REDUCE_BLOCK_N = 128
            with torch.cuda.stream(stream):
                grid = lambda META: (
                    triton.cdiv(M, META["BLOCK_SIZE_M"]) * triton.cdiv(N, META["BLOCK_SIZE_N"]),
                )
                gemm_rs_nvlink_kernel[grid](
                    x,
                    self.weight,
                    self.output_tensor_ptrs,
                    M,
                    N,
                    K_per_rank,
                    blocks_m_per_rank,
                    x.stride(0),
                    x.stride(1),
                    self.weight.stride(0),
                    self.weight.stride(1),
                    self.output_tensor.stride(0),
                    self.output_tensor.stride(1),
                    self.rank,
                    self.world_size,
                    self.barrier_ptrs,
                    GEMM_BLOCK_M,
                    GEMM_BLOCK_N,
                    GEMM_BLOCK_K,
                    GEMM_GROUP_M_SIZE,
                    num_stages=4,
                    num_warps=4,
                )

            with torch.cuda.stream(self.rs_stream):
                grid = lambda META: (
                    triton.cdiv(m_per_rank, META["BLOCK_M"]) * triton.cdiv(N, META["BLOCK_N"]),
                )
                reduce_nvlink_kernel[grid](
                    self.output_tensor,
                    self.output,
                    M,
                    N,
                    self.output.stride(0),
                    self.output.stride(1),
                    self.rank,
                    self.world_size,
                    self.barrier,
                    REDUCE_BLOCK_M,
                    REDUCE_BLOCK_N,
                    GEMM_BLOCK_M,
                    GEMM_BLOCK_N,
                )
            stream.wait_stream(self.rs_stream)
        if self.dump_ptx:
            if self.rank == 0:
                if self.fused:
                    raise NotImplementedError()
                else:
                    with open("debug_output_GEMM_RS_NVLink_Triton.ptx", "w") as a:
                        print(list(gemm_rs_nvlink_kernel.cache[0].values())[0].asm["ptx"], file=a)
        self.group_barrier.barrier_all(stream.cuda_stream)
        return self.output[:m_per_rank, :]

File: src/pybind/gemm_rs.cc
//===- gemm_rs.cc ------------------------------------------------- C++ ---===//
//
// Copyright 2025 ByteDance Ltd. and/or its affiliates. All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//===----------------------------------------------------------------------===//

#include "flux/ths_op/flux_shm.h"
#include "gemm_rs/ths_op/gemm_reduce_scatter.h"
#include "gemm_rs/ths_op/helper_ops.h"
#include "flux/ths_op/ths_pybind.h"
#include "gemm_rs/tile_scheduler/threadblock_swizzle_segment_util.hpp"

namespace bytedance::flux::ths_op {
using GemmRSOpCls = TorchClassWrapper<GemmRS>;
namespace py = pybind11;

static int _register_gemm_rs_ops [[maybe_unused]] = []() {
  ThsOpsInitRegistry::instance().register_one("gemm_reduce_scatter", [](py::module &m) {
    py::class_<GemmRSOpCls>(m, "GemmRS")
        .def(
            py::init([](c10::intrusive_ptr<c10d::ProcessGroup> tp_group,
                        int32_t nnodes,
                        int32_t max_m,
                        int32_t n_dim,
                        torch::ScalarType input_dtype,
                        py::object py_output_dtype,
                        bool transpose_weight,
                        bool fuse_reduction,
                        bool ring_reduction) {
              auto output_dtype = py_output_dtype.is(py::none())
                                      ? input_dtype
                                      : torch::python::detail::py_object_to_dtype(py_output_dtype);

              return new GemmRSOpCls(
                  std::make_shared<C10dProcessGroup>("", tp_group),
                  nnodes,
                  max_m,
                  n_dim,
                  input_dtype,
                  output_dtype,
                  transpose_weight,
                  fuse_reduction,
                  ring_reduction);
            }),
            py::arg("tp_group"),
            py::arg("nnodes"),
            py::arg("max_m"),
            py::arg("n_dim"),
            py::arg("input_dtype"),
            py::arg("output_dtype") = py::none(),
            py::arg("transpose_weight") = false,
            py::arg("fuse_reduction") = false,
            py::arg("ring_reduction") = false)
        .def("zero_buffers", &GemmRSOpCls::zero_buffers)
        .def(
            "forward",
            &GemmRSOpCls::forward,
            py::arg("input"),
            py::arg("weight"),
            py::arg("bias") = py::none(),
            py::arg("input_scale") = py::none(),
            py::arg("weight_scale") = py::none(),
            py::arg("output_scale") = py::none(),
            py::arg("fast_accum") = false,
            py::arg("reduce_scatter_option") = ReduceScatterOptionWithOptional())
        .def(
            "forward_barrier",
            &GemmRSOpCls::forward_barrier,
            py::arg("input"),
            py::arg("weight"),
            py::arg("bias") = py::none())
        .def(
            "forward_reduce_scatter",
            &GemmRSOpCls::forward_reduce_scatter,
            py::arg("input"),
            py::arg("weight"),
            py::arg("bias") = py::none(),
            py::arg("input_scale") = py::none(),
            py::arg("weight_scale") = py::none())
        .def(
            "profiling",
            &GemmRSOpCls::profiling,
            py::arg("input"),
            py::arg("weight"),
            py::arg("bias") = py::none(),
            py::arg("input_scale") = py::none(),
            py::arg("weight_scale") = py::none(),
            py::arg("output_scale") = py::none(),
            py::arg("fast_accum") = false,
            py::arg("prof_ctx") = nullptr,
            py::arg("reduce_scatter_option") = ReduceScatterOptionWithOptional());

    py::class_<SegmentInfo>(m, "SegmentInfo")
        .def_readonly("segment_origin", &SegmentInfo::segment_origin)
        .def_readonly("size", &SegmentInfo::size)
        .def_readonly("tile_m_start_origin", &SegmentInfo::tile_m_start_origin)
        .def_readonly("tile_m_start_new", &SegmentInfo::tile_m_start_new);

    m.def(
        "get_gemm_rs_threadblock_segments_info",
        [](std::vector<int> problem_shape,  // (M, N)
           std::vector<int> tiled_shape,
           int rank,
           int world_size,
           int sub_world_size,
           int nnodes,
           bool use_2d_ring,
           bool per_tile_flags) {
          FLUX_CHECK(problem_shape.size() == 2 || problem_shape.size() == 3);
          FLUX_CHECK(tiled_shape.size() == 2 || tiled_shape.size() == 3);
          ThreadBlockSwizzleSegmentUtils helper(
              cutlass::gemm::GemmCoord(problem_shape[0], problem_shape[1], 1),
              cutlass::gemm::GemmCoord(tiled_shape[0], tiled_shape[1], 1),
              rank,
              world_size,
              sub_world_size,
              nnodes,
              use_2d_ring,
              per_tile_flags);
          return helper.get_segments_info();
        },
        py::arg("problem_shape"),
        py::arg("tiled_shape"),
        py::arg("rank"),
        py::arg("world_size"),
        py::arg("sub_world_size"),
        py::arg("nnodes"),
        py::arg("use_2d_ring"),
        py::arg("per_tile_flags"));
    m.def(
        "calc_gemm_rs_threadblock_segments_info",
        [](torch::Tensor segments,
           std::vector<int> problem_shape,  // (M, N)
           std::vector<int> tiled_shape,    // (TILED_M, TILED_N)
           int rank,
           int world_size,
           int sub_world_size,
           int nnodes,
           bool use_2d_ring,
           bool per_tile_flags) {
          FLUX_CHECK(problem_shape.size() == 2 || problem_shape.size() == 3);
          FLUX_CHECK(tiled_shape.size() == 2 || tiled_shape.size() == 3);
          ThreadBlockSwizzleSegmentUtils helper(
              cutlass::gemm::GemmCoord(problem_shape[0], problem_shape[1], 1),
              cutlass::gemm::GemmCoord(tiled_shape[0], tiled_shape[1], 1),
              rank,
              world_size,
              sub_world_size,
              nnodes,
              use_2d_ring,
              per_tile_flags);
          return helper.calc_segments_info((SegmentInfo *)segments.data_ptr());
        },
        py::arg("segments"),
        py::arg("problem_shape"),
        py::arg("tiled_shape"),
        py::arg("rank"),
        py::arg("world_size"),
        py::arg("sub_world_size"),
        py::arg("nnodes"),
        py::arg("use_2d_ring"),
        py::arg("per_tile_flags"));
    m.def("bsr_reduce", &bsr_reduce);
    m.def("pad_m_to_TPxTile", &pad_m_to_TPxTile);
  });
  return 0;
}();
}  // namespace bytedance::flux::ths_op
